MobiusCore/Source/EffectHandlers/EffectExecutor.swift

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2019 Spotify AB.
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+
+class EffectExecutor<Effect, Event>: Connectable {
+    private let handleEffect: (Effect, Response<Event>) -> Disposable
+    private var output: Consumer<Event>?
+
+    private let lock = Lock()
+
+    // Keep track of each received effect's state.
+    // When an effect has completed, it should be removed from this dictionary.
+    // When disposing this effect handler, all entries must be removed.
+    private var handlingEffects: [Int64: EffectHandlingState<Event>] = [:]
+    private var nextID = Int64(0)
+
+    init(handleInput: @escaping (Effect, Response<Event>) -> Disposable) {
+        self.handleEffect = handleInput
+    }
```

Comment:
Any `Response`s that were still inflight would be ended, which would do nothing since they only hold a weak reference to `EffectExecutor` objects, and all the work they do in `end` is remove the references held by `EffectExecutor`.

Do you think we should be disposing in `deinit`?

Reply:
As a general rule, I think disposables should either dispose or assert that they’re already disposed in deinit.